### PR TITLE
Fix `gather` for sliced input structs column

### LINKED ISSUE #9213: [BUG] `lists::detail::sort_lists` returns wrong result for sliced lists column containing structs entries
During working on `drop_list_duplicates` for structs entries, a bug shows up indicating that sliced columns of structs are not properly handled in `lists::detail::sort_lists`:
 * With `cudf::test::print(lists_original)`:
```
List<Struct<int8_t,cudf::string_view,>>:
Length : 3
Offsets : 0, 8, 16, 24
   Struct<int8_t,cudf::string_view,>:
   Length : 24:
      1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 2, 2, 2, 2, 2, 2, 3, 2, 3, 3
      Banana, Mango, Apple, Cherry, Kiwi, Banana, Cherry, Kiwi, Bear, Duck, Cat, Dog, Panda, Bear, Cat, Panda, ÁÁÁ, ÉÉÉÉÉ, ÍÍÍÍÍ, ÁBC, XYZ, ÁÁÁ, ÁBC, XYZ
```

 * With `auto const lists    = cudf::slice(lists_original, {1, 3})[0];`:
```
List<Struct<int8_t,cudf::string_view,>>(sliced):
Length : 2
Offsets : 0, 8, 16
   Struct<int8_t,cudf::string_view,>:
   Length : 16:
      1, 1, 1, 1, 2, 1, 2, 2, 2, 2, 2, 2, 3, 2, 3, 3
      Bear, Duck, Cat, Dog, Panda, Bear, Cat, Panda, ÁÁÁ, ÉÉÉÉÉ, ÍÍÍÍÍ, ÁBC, XYZ, ÁÁÁ, ÁBC, XYZ
```

 * When calling `detail::sort_lists(lists_column_view(lists), order::ASCENDING, null_order::AFTER);`:
```
List<Struct<int8_t,cudf::string_view,>>:
Length : 2
Offsets : 0, 8, 16
   Struct<int8_t,cudf::string_view,>:
   Length : 16:
      1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2
      Banana, Banana, Apple, Cherry, Mango, Cherry, Kiwi, Kiwi, Dog, Bear, Bear, Duck, Cat, Panda, Panda, Cat
```

Note that `sort_lists` works just fine with sliced columns of basic types. I quickly look at the implementation of `sort_lists` and see that it calls `segmented_sorted_order` and `gather` so maybe the bug is related to one of these APIs.

### PR DESCRIPTION
**Pull request by ttnghia**
This PR fixes the `gather` API for structs columns when the input is a sliced column. Previously, `gather` calls `child_begin()` and `child_end()` to access the children column so if the input structs column is sliced then the output is incorrect.

This closes #9213, and is blocked by https://github.com/rapidsai/cudf/pull/9194 due to conflict work.

### CODE REVIEW

**Comment by ttnghia** [comment]
Rerun tests.

File: cpp/include/cudf/detail/gather.cuh [hunk 1]
```diff
@@ -450,16 +450,24 @@ struct column_gatherer_impl<struct_view> {
...
+                   thrust::make_counting_iterator(column.num_children()),
+                   std::back_inserter(sliced_children),
+                   [structs_view = structs_column_view{column}](auto const idx) {
+                     return structs_view.get_sliced_child(idx);
+                   });
+
     std::vector<std::unique_ptr<cudf::column>> output_struct_members;
-    std::transform(structs_column.child_begin(),
-                   structs_column.child_end(),
+    std::transform(sliced_children.begin(),
```
Comment (mythrocks) [inline cpp/include/cudf/detail/gather.cuh:467]: This is a good catch. :/

I really think `struct_view` should have a `sliced_child_begin()`, or some such. I suspect this pattern will repeat across code that needs to support `STRUCT`. A later PR, maybe.

**Comment by ttnghia** [comment]
Rerun tests.

**Review by mythrocks — APPROVED** [review]

**Comment by mythrocks** [comment]
Hey, @ttnghia. I just merged #9194. Please rebase this change.

**Review by harrism — APPROVED** [review]

**Comment by ttnghia** [comment]
@gpucibot merge

**Pull request merged**